File: py/handout-01/postfixparser.py
class PostfixParser:
  """Parses, validates, and and evaluates a postfix expression"""
  
  # the values that the program accepts
  values = {
    'a': 5,
    'b': 7,
    'c': 2,
    'd': 4
  }
  # the set of all valid inputs
  valid_inputs = set(list(values.keys()) + ['+', '-', '*', '/'])

  def __eval_binary_expr(self, first: float, second: float, op: str) -> float:
    """Evaluate a binary expression given two ordered values and an operator and 
    return the result
    
    :param first: the first operand
    :type first: float
    :param second: the second operand
    :type second: float
    :param op: the operator
    :type op: str
    :returns: the result of the expression
    :rtype: float
    """

    if op == '+':
      return first + second
    if op == '-':
      return first - second
    if op == '*':
      return first * second
    if op == '/':
      return first / second
    raise Exception(f'Invalid operator \'{op}\'')

  def eval_postfix_expr(self, expr: str) -> float:
    """Parse a postfix expression and return the result
    
    :param expr: the expression to evaluate
    :type expr: str
    :returns: the result of the expression
    :rtype: float
    """

    if len(expr) == 0:
      raise Exception('Empty expression')

    stack = []
    for char in expr:
      # validate character
      if char not in self.valid_inputs:
        raise Exception(f'Valid expression inputs: {self.valid_inputs}')
      # check if current character is a number
      if char in self.values:
        stack.append(self.values[char])
      # current character is an operator
      else:
        # pop last two elements from the stack
        if len(stack) < 2:
          raise Exception('Invalid expression')
        last, second_last = stack.pop(), stack.pop()
        # evaluate the binary expression with the current operator
        result = self.__eval_binary_expr(
          first=second_last, op=char, second=last
        )
        # push the result onto the stack
        stack.append(result)
    
    # there should be exactly one element left in the stack or the expression 
    # was invalid
    if len(stack) != 1:
      raise Exception('Invalid expression')
    return stack[0]

File: cxx/handout-02/postfixparser.cpp
#include "postfixparser.h"

bool PostfixParser::isInt(const std::string& s) {
  // check if each character is a digit
  for (const char& c : s) {
    if (!std::isdigit(c)) {
      return false;
    }
  }
  return true;
}

void PostfixParser::getTokens(const std::string& expr) {
  bool done;
  std::string token;

  int i = 0, n = expr.length();
  done = false;
  // split the expression into a vector of individual tokens
  while (!done) {
    // ignore leading whitespace
    token.clear();
    while (i < n && std::isspace(expr[i])) {
      ++i;
    }
    // read characters, appending to the current token and then pushing it back
    // to the vector once the end of expression or another space is encountered
    while (i < n) {
      if (expr[i] == '$') {
        // done, push back last token read and exit loops
        if (token.length() > 0) {
          this->tokens.push_back(token);
        }
        done = true;
        break;
      }
      if (!std::isspace(expr[i])) {
        token += expr[i];
        ++i;
      } else {
        this->tokens.push_back(token);
        ++i;
        break;
      }
    }
  }
}

void PostfixParser::getVars() {
  std::string value;

  // iterate over tokens
  for (const std::string& tok : this->tokens) {
    // if the current token is an identifier, read in its value and add it to
    // the vars map
    if (std::isalpha(tok[0])) {
      std::cout << "Enter the value of " << tok << ": ";
      std::getline(std::cin, value);
      if (!isInt(value)) {
        throw std::runtime_error("Expected integer, received {" + value + '}');
      }
      this->vars[tok] = std::stoi(value);
    }
  }
}

int PostfixParser::evalBinaryExpr(int l, char op, int r) const {
  switch (op) {
  case '+': return l + r;
  case '-': return l - r;
  case '*': return l * r;
  case '/': return l / r;
  default: throw std::runtime_error("Expected operator, received {" + op + '}');
  }
}

int PostfixParser::evalExpr() const {
  int l, r;
  std::stack<int> stack;

  // iterate over tokens
  for (const std::string& token : this->tokens) {
    if (this->vars.find(token) != this->vars.end()) { // token is identifier
      stack.push(this->vars.at(token));
    } else if (isInt(token)) { // token is digit
      stack.push(stoi(token));
    } else { // token is operator
      // validate operator and current state of the stack
      if (token != "+" && token != "-" && token != "*" && token != "/") {
        throw std::runtime_error("Expected operator, received {" + token + '}');
      }
      if (stack.size() < 2) {
        throw std::runtime_error("Expected two operands for operator " + token);
      }
      // evaluate expression and push result onto stack
      r = stack.top();
      stack.pop();
      l = stack.top();
      stack.pop();
      stack.push(evalBinaryExpr(l, token[0], r));
    }
  }

  // should be exactly one element left in the stack or else the expression is
  // invalid
  if (stack.size() != 1) {
    throw std::runtime_error("Invalid expression");
  }

  return stack.top();
}